Code/Engine/GameEngine/Gameplay/Implementation/TriggerDelayModifierComponent.cpp

```diff
@@ -0,0 +1,158 @@
+#include <GameEngine/GameEnginePCH.h>
+
+#include <Core/Messages/TriggerMessage.h>
+#include <Core/World/GameObject.h>
+#include <Core/WorldSerializer/WorldReader.h>
+#include <Core/WorldSerializer/WorldWriter.h>
+#include <GameEngine/Gameplay/TriggerDelayModifierComponent.h>
+
+// clang-format off
+EZ_BEGIN_COMPONENT_TYPE(ezTriggerDelayModifierComponent, 1 /* version */, ezComponentMode::Static)
+{
+  EZ_BEGIN_PROPERTIES
+  {
+    EZ_MEMBER_PROPERTY("ActivationDelay", m_ActivationDelay),
+    EZ_MEMBER_PROPERTY("DeactivationDelay", m_DeactivationDelay),
+  }
+  EZ_END_PROPERTIES;
+
+  EZ_BEGIN_MESSAGEHANDLERS
+  {
+    EZ_MESSAGE_HANDLER(ezMsgComponentInternalTrigger, OnMsgComponentInternalTrigger),
+    EZ_MESSAGE_HANDLER(ezMsgTriggerTriggered, OnMsgTriggerTriggered),
+  }
+  EZ_END_MESSAGEHANDLERS;
+
+  EZ_BEGIN_ATTRIBUTES
+  {
+    new ezCategoryAttribute("Gameplay/Logic"), // Component menu group
+  }
+  EZ_END_ATTRIBUTES;
+}
+EZ_END_COMPONENT_TYPE
+// clang-format on
+
+ezTriggerDelayModifierComponent::ezTriggerDelayModifierComponent() = default;
+ezTriggerDelayModifierComponent::~ezTriggerDelayModifierComponent() = default;
+
+void ezTriggerDelayModifierComponent::Initialize()
+{
+  SUPER::Initialize();
+}
+
+void ezTriggerDelayModifierComponent::SerializeComponent(ezWorldWriter& stream) const
+{
+  SUPER::SerializeComponent(stream);
+  auto& s = stream.GetStream();
+
+  s << m_ActivationDelay;
+  s << m_DeactivationDelay;
+}
+
+void ezTriggerDelayModifierComponent::DeserializeComponent(ezWorldReader& stream)
+{
+  SUPER::DeserializeComponent(stream);
+  // const ezUInt32 uiVersion = stream.GetComponentTypeVersion(GetStaticRTTI());
+  auto& s = stream.GetStream();
+
+  s >> m_ActivationDelay;
+  s >> m_DeactivationDelay;
+}
+
+void ezTriggerDelayModifierComponent::OnMsgTriggerTriggered(ezMsgTriggerTriggered& msg)
+{
+  if (msg.m_TriggerState == ezTriggerState::Activated)
+  {
+    if (m_iElementsInside++ == 0) // was 0 before the increment
+    {
+      // the first object entered the trigger
+
+      if (!m_bIsActivated)
+      {
+        // the trigger is not active yet -> send an activation message with a new activation token
+
+        ++m_iValidActivationToken;
+
+        // store the original trigger message for later
+        m_sMessage = msg.m_sMessage;
```

Comment:
msg.m_hTriggeringObject is lost

Reply:
Yes, because there is no 1:1 relationship anymore who entered/left a trigger.

Reply:
ah yes makes sense